src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -430,28 +449,71 @@ private static ParquetBufferValue getStructValue(
       SubColumnFinder subColumnFinder,
       ZoneId defaultTimezone,
       final long insertRowsCurrIndex,
-      boolean isDescendantsOfRepeatingGroup) {
+      boolean isDescendantsOfRepeatingGroup,
+      InsertValidationResponse.InsertError error) {
     Map<String, ?> structVal =
         DataValidationUtil.validateAndParseIcebergStruct(
             subColumnFinder.getDotPath(type.getId().toString()), value, insertRowsCurrIndex);
     Set<String> extraFields = new HashSet<>(structVal.keySet());
+    List<String> missingFields = new ArrayList<>();
     List<Object> listVal = new ArrayList<>(type.getFieldCount());
     float estimatedParquetSize = 0f;
     for (int i = 0; i < type.getFieldCount(); i++) {
-      ParquetBufferValue parsedValue =
-          parseColumnValueToParquet(
-              structVal.getOrDefault(type.getFieldName(i), null),
-              type.getType(i),
-              statsMap,
-              subColumnFinder,
-              defaultTimezone,
-              insertRowsCurrIndex,
-              isDescendantsOfRepeatingGroup);
       extraFields.remove(type.getFieldName(i));
-      listVal.add(parsedValue.getValue());
-      estimatedParquetSize += parsedValue.getSize();
+      if (structVal.containsKey(type.getFieldName(i))) {
+        ParquetBufferValue parsedValue =
+            parseColumnValueToParquet(
+                structVal.get(type.getFieldName(i)),
+                type.getType(i),
+                statsMap,
+                subColumnFinder,
+                defaultTimezone,
+                insertRowsCurrIndex,
+                isDescendantsOfRepeatingGroup,
+                error);
+        listVal.add(parsedValue.getValue());
+        estimatedParquetSize += parsedValue.getSize();
+      } else {
+        if (type.getType(i).isRepetition(Repetition.REQUIRED)) {
+          missingFields.add(type.getFieldName(i));
+        } else {
+          listVal.add(null);
+        }
+      }
+    }
+
+    if (!missingFields.isEmpty()) {
+      missingFields.forEach(
+          missingField -> {
+            if (error != null) {
+              List<String> missingFieldPath =
+                  new ArrayList<>(subColumnFinder.getPath(type.getId().toString()));
+              missingFieldPath.add(missingField);
+              error.addMissingNotNullColName(
+                  concatDotPath(missingFieldPath.toArray(new String[0])));
+            }
+          });
+      String missingFieldsStr = missingFields.stream().collect(Collectors.joining(", ", "[", "]"));
+      throw new SFException(
```

Comment:
left this comment in another place too - we should not be throwing exceptions on extra cols / missing cols / unexpected-null-value until all of these errors have been collected. Right now we are throwing on the first possible error which is a deviation from current behavior and makes KC's life harder.